tfjs-converter/src/executor/model_analysis.ts

```diff
@@ -146,26 +145,87 @@ export function getNodesInTopologicalOrder(
   return orderedNodes;
 }
 
-const CONTROL_FLOW_OPS = [
+/**
+ * Given the execution info, return a map from node to the disposable node list
+ * after its execution.
+ *
+ * @returns A map from node to disposable nodes after its
+ *     execution. That is, for a node `x`, `nodeLiveUntilMap[x]` indicates all
+ *     nodes which their intermediate tensors should be disposed after `x` being
+ *     executed.
+ */
+export function getNodeLiveUntilMap(orderedNodes: Node[]): Map<Node, Node[]> {
+  const nNodes = orderedNodes.length;
+  const nodeToOrder = new Map<Node, number>();
+  for (let i = 0; i < nNodes; ++i) {
+    nodeToOrder.set(orderedNodes[i], i);
+  }
+
+  // `liveUntil[i]` indicates that "all the intermediate tensors from
+  // `orderedNodes[i]` should be disposed after `orderedNodes[liveUntil[i]]`
+  // being executed."
+  const INF_LIFE = Math.floor(Number.MAX_SAFE_INTEGER / 2);
+  const liveUntil = [...Array(nNodes).keys()];
+
+  for (let nodeOrder = 0; nodeOrder < nNodes; ++nodeOrder) {
+    const node = orderedNodes[nodeOrder];
+    // Skip any control flow nodes, since its dependency is tricky to track
+    // correctly.
+    if (isControlFlow(node)) {
+      liveUntil[nodeOrder] = INF_LIFE;
+    }
+  }
+
+  for (let nodeOrder = 0; nodeOrder < nNodes; ++nodeOrder) {
+    const node = orderedNodes[nodeOrder];
+    for (const child of node.children) {
+      const childOrder = nodeToOrder.get(child)!;
+      // Extend the node's life to at least its child's life.
+      liveUntil[nodeOrder] =
+          Math.max(liveUntil[nodeOrder], liveUntil[childOrder]);
+    }
+  }
```

Comment:
Nit: This may be easier to read, but feel free to skip it.

```suggestion
  const liveUntil = orderedNodes.map((node, index) => {
    if (isControlFlow(node)) {
      return INF_LIFE;
    }
    return node.children
          .map(node => nodeToOrder.get(node)!)
          .reduce((a, b) => Math.max(a, b), index);
  });
```

Reply:
Just notice that this rewrite this incorrect. If a child node lives until INF_LIFE, the parent node should also live until INF_LIFE. That's for the control flow node so that it can always get tenors from its parents.

This is also the behavior of the old `checkNodesForDisposal` function, as far as I understand. Please double-check this if you can.

Reply:
Nice catch. Like you said, I missed the `INF_LIFE` of children also affecting parents. I think this fixes it:
```ts
  // Make control flow nodes (and consequently their direct parents)
  // live forever since they're tricky to track correctly.
  const selfLifespans = orderedNodes.map((node, nodeOrder) => 
    isControlFlow(node) ? INF_LIFE : nodeOrder);

  // A node lives long enough to pass on its tensors to its children.
  // It lives until at least `max(node's position, childrens' positions)`.
  const liveUntil = orderedNodes.map((node, nodeOrder) =>
    node.children
      .map(node => selfLifespans[nodeToOrder.get(node)!])
      .reduce((a, b) => Math.max(a, b), selfLifespans[nodeOrder]));
```

Reply:
Nice. I didn't know Math.max uses recursion in it, thanks!